File: backend/app/schemas/blog.py
"""Request/response bodies for blog CMS APIs."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BlogPostCreate(BaseModel):
    slug: str = Field(..., min_length=1, max_length=160)
    title: str = Field(..., min_length=1, max_length=300)
    meta_title: str | None = Field(None, max_length=320)
    meta_description: str = Field(..., min_length=1)
    body: str = Field(..., min_length=1)
    status: str = Field(default="draft", pattern="^(draft|scheduled|published|archived)$")
    scheduled_at: datetime | None = None
    cover_image_url: str | None = None
    category_id: str | None = None
    tag_ids: list[str] = Field(default_factory=list)
    og_title: str | None = Field(None, max_length=300)
    og_description: str | None = None
    og_image_url: str | None = None
    canonical_url: str | None = Field(None, max_length=500)
    robots_directives: str | None = Field(None, max_length=120)
    keywords: str | None = Field(None, max_length=500)
    schema_jsonld: dict | None = None


class BlogPostPublicListItem(BaseModel):
    """Published posts index (marketing site)."""

    id: str
    slug: str
    title: str
    meta_description: str
    published_at: str | None
    updated_at: str | None


class BlogPostAdminListItem(BaseModel):
    """Admin post list rows (all statuses)."""

    id: str
    slug: str
    title: str
    meta_description: str
    status: str
    published_at: str | None
    updated_at: str | None


class BlogPostDetail(BaseModel):
    """Full post payload returned by public and admin detail endpoints."""

    id: str
    slug: str
    title: str
    meta_title: str | None
    meta_description: str
    body: str
    og_title: str | None
    og_description: str | None
    og_image_url: str | None
    canonical_url: str | None
    robots_directives: str | None
    keywords: str | None
    cover_image_url: str | None
    schema_jsonld: dict | None
    status: str
    scheduled_at: str | None
    published_at: str | None
    updated_at: str | None

File: frontend/src/lib/blog.ts
import 'server-only';

import { getInternalApiBase } from '@/lib/internal-api';

export type BlogListItem = {
  id: string;
  slug: string;
  title: string;
  meta_description: string;
  published_at: string | null;
  updated_at: string | null;
};

export type BlogPostPublic = BlogListItem & {
  meta_title: string | null;
  body: string;
  og_title: string | null;
  og_description: string | null;
  og_image_url: string | null;
  canonical_url: string | null;
  robots_directives: string | null;
  keywords: string | null;
  schema_jsonld: Record<string, unknown> | null;
};

export async function fetchPublishedPosts(): Promise<BlogListItem[]> {
  const res = await fetch(`${getInternalApiBase()}/blog/posts`, {
    next: { revalidate: 60 },
  });
  if (!res.ok) {
    return [];
  }
  return (await res.json()) as BlogListItem[];
}

export async function fetchPublishedPostBySlug(
  slug: string,
): Promise<BlogPostPublic | null> {
  const res = await fetch(
    `${getInternalApiBase()}/blog/posts/${encodeURIComponent(slug)}`,
    { next: { revalidate: 60 } },
  );
  if (res.status === 404) {
    return null;
  }
  if (!res.ok) {
    return null;
  }
  return (await res.json()) as BlogPostPublic;
}

export async function fetchPublishedSlugList(): Promise<string[]> {
  const posts = await fetchPublishedPosts();
  return posts.map((p) => p.slug);
}
